src/bt.rs

```diff
@@ -0,0 +1,423 @@
+//! tree.insert(bits!(1, 1), 'f');
+
+use core::fmt::Debug;
+use std::io::Cursor;
+
+use bitvec::slice::BitSlice;
+
+use crate::codec::{CodecError, Decode, Encode, ParameterizedDecode};
+
+/// Errors triggered by binary tree operations.
+#[derive(Debug, thiserror::Error)]
+#[non_exhaustive]
+pub enum BinaryTreeError<V> {
+    /// Error when inserting in a node with existing value.
+    #[error("node already contains a value")]
+    InsertNonEmptyNode(V),
+    /// Error when an operation cannot reach the node at the end of a path.
+    #[error("unreachable node in the tree")]
+    UnreachableNode(V),
+}
+
+/// Used to indicate a traversal path on the binary tree.
+pub type Path = BitSlice;
+
+type SubTree<V> = Option<Box<Node<V>>>;
+
+/// Represents a node of a binary tree.
+pub struct Node<V> {
+    value: V,
+    left: SubTree<V>,
+    right: SubTree<V>,
+}
+
+impl<V> Node<V> {
+    fn new(value: V) -> Self {
+        Self {
+            value,
+            left: None,
+            right: None,
+        }
+    }
+
+    /// Inserts the value at the end of the path.
+    ///
+    /// # Returns
+    /// - `Ok(())`, the node was inserted at the end of the path.
+    /// - `Err(InsertNonEmptyNode(value))`, the node already contains a value.
+    /// - `Err(UnreachablePath(value))`, the end of the path is unreachable.
+    ///
+    /// In the error cases, no insertion occurs and the value is returned to its owner.
+    pub fn insert(&mut self, path: &Path, value: V) -> Result<(), BinaryTreeError<V>> {
+        let mut iter = path.iter();
+
+        if let Some(first_bit) = iter.next() {
+            let mut node = if !first_bit {
+                &mut self.left
+            } else {
+                &mut self.right
+            };
+
+            // Finds the node at the end of the path.
+            for bit in iter {
+                if let Some(next_node) = node {
+                    node = if !bit {
+                        &mut next_node.left
+                    } else {
+                        &mut next_node.right
+                    };
+                } else {
+                    return Err(BinaryTreeError::UnreachableNode(value));
+                }
+            }
+
+            // Checks whether the node already has a value,
+            // If so, returns an error wrapping the value to be inserted.
+            // otherwise, node is available for insertion, and `Ok(())` is returned.
+            if node.is_some() {
+                Err(BinaryTreeError::InsertNonEmptyNode(value))
+            } else {
+                *node = Some(Box::new(Node::new(value)));
+                Ok(())
+            }
+        } else {
+            Err(BinaryTreeError::InsertNonEmptyNode(value))
+        }
+    }
+
+    /// Gets a reference to the value located at the end of the path.
+    /// [`None`] is returned if no values are found during the traversal of the tree
+    /// along the path.
+    pub fn get(&self, path: &Path) -> Option<&V> {
```

Comment:
Could method use `get_node()` as a subroutine?

Reply:
unfortunately not, due to mutability.